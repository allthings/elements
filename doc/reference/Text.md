<!-- 
This is an auto-generated markdown. 
You can change it in "/Users/daniel/Dev/allthings/elements/src/Text/Text.tsx" and run build:docs to update this file.
-->
# createTextStyles

## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|size|"xs", "s", "m", "l", "xl", "xxl", "giant"|
|block|boolean|
|italic|boolean|
|strong|boolean|
|underline|boolean|
|lineThrough|boolean|
|align|"left", "center", "right"|
|autoBreak|boolean|
