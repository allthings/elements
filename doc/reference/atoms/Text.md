# Text
Text will be used for everywhere a text appears.
The only exception is in molecules that already provide the
text component for you.

```example
<Text>
  Text will be used for everywhere a text appears. The only exception
  is in molecules that already provide the text component for you.
</Text>
```

```example
<Text autoBreak={true}>
  Textwillbeusedforeverywhereatextappears.Theonlyexceptionisinmoleculesthatalreadyprovidethetextcomponentforyou.Textwillbeusedforeverywhereatextappears.Theonlyexceptionisinmoleculesthatalreadyprovidethetextcomponentforyou.
 </Text>
## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|align|enum|
|autoBreak|bool|
|block|bool|<br>Default: true
|children|node|
|color|string|<br>Default: 'text'
|italic|bool|
|size|enum|<br>Default: 'l'
|strong|bool|
|underline|bool|
|lineThrough|bool|
