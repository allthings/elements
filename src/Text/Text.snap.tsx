// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text renders correctly 1`] = `
.css-0,
[data-css-0] {
  display: block;
  font-family: "Open Sans", Helvetica, Arial, sans-serif;
  font-size: 14px;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: #333333;
}

.css-1,
[data-css-1] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

<div
  data-css-0=""
  data-css-1=""
/>
`
