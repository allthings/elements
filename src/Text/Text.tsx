import { css } from 'glamor'
import neue from 'neue'
import React from 'react'
import { useTheme } from '../Theme'
import View, { IViewProps } from '../View'

if (typeof window !== `undefined`) {
  neue.load([
    {
      families: ['Open Sans:n3,n3i,n4,n4i,n6,n6i,n7,n7i'],
      css:
        '//fonts.googleapis.com/css?family=Open+Sans:300,300i,400,400i,600,600i,700,700i',
    },
  ])
}

const availableSizes = {
  xs: 10,
  s: 12,
  m: 13,
  l: 14,
  xl: 18,
  xxl: 18,
  giant: 24,
}

export type TextSizeType = number | keyof typeof availableSizes

const textWeightMap = {
  light: 300,
  regular: 400,
  'semi-bold': 600,
  bold: 700,
}

export type TextWeightType = keyof typeof textWeightMap

interface ITextStyles {
  size?: TextSizeType
  block?: boolean
  italic?: boolean
  /**
   * @deprecated
   */
  strong?: boolean
  underline?: boolean
  lineThrough?: boolean
  weight?: TextWeightType
  align?: 'left' | 'center' | 'right'
  autoBreak?: boolean
}

export const createTextStyles = ({
  block = false,
  italic = false,
  strong = false,
  size = 'l',
  weight,
  underline = false,
  lineThrough = false,
  align,
  autoBreak,
}: ITextStyles = {}) =>
  css({
    display: block ? 'block' : 'inline',
    fontFamily: '"Open Sans", Helvetica, Arial, sans-serif',
    fontStyle: italic && 'italic',
    fontWeight: weight ? textWeightMap[weight] : strong ? 600 : 400,
    fontSize: typeof size === 'number' ? size : availableSizes[size],
    textDecoration:
      (underline && 'underline') || (lineThrough && 'line-through'),
    textAlign: align,
    whiteSpace: autoBreak && 'pre-wrap',
    wordBreak: autoBreak && 'break-word',
    WebkitFontSmoothing: 'antialiased',
    MozOsxFontSmoothing: 'grayscale',
  })

export interface IText {
  readonly color?: string
}

export type ITextProps = IText & ITextStyles & IViewProps

let hasWarnedBefore = false

/**
 * Text will be used for everywhere a text appears.
 * The only exception is in molecules that already provide the
 * text component for you.
 *
 * ```example
 * <Text>
 *   Text will be used for everywhere a text appears. The only exception
 *   is in molecules that already provide the text component for you.
 * </Text>
 * ```
 *
 * ```example
 * <Text autoBreak={true}>
 *   Textwillbeusedforeverywhereatextappears.Theonlyexceptionisinmoleculesthatalreadyprovidethetextcomponentforyou.Textwillbeusedforeverywhereatextappears.Theonlyexceptionisinmoleculesthatalreadyprovidethetextcomponentforyou.
 *  </Text>
 *  ```
 */
const Text = ({
  color = 'text',
  block = true,
  children,
  italic,
  size = 'l',
  strong,
  align,
  underline,
  autoBreak,
  lineThrough,
  weight,
  ...props
}: ITextProps) => {
  const { colorize } = useTheme()

  if (strong === true && process.env.NODE_ENV !== 'production') {
    !hasWarnedBefore &&
      console.warn(
        "The property `strong` is deprecated. Please use `weight` instead. (`strong` now corresponds to weight: 'semi-bold')",
      )
    hasWarnedBefore = true
  }

  return (
    <View
      {...css(
        createTextStyles({
          block,
          italic,
          strong,
          size,
          underline,
          lineThrough,
          align,
          autoBreak,
          weight,
        }),
        { color: colorize(color) },
      )}
      {...props}
    >
      {children}
    </View>
  )
}

export default Text
