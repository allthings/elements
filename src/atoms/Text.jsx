import PropTypes from 'prop-types'
import React from 'react'
import { css } from 'glamor'
import View from '../atoms/View'
import Theme from '../behaviour/Theme'
import WebFont from 'webfontloader'

if (typeof window !== `undefined`) {
  WebFont.load({
    classes: false,
    google: {
      families: ['Open Sans:300,600'],
    },
  })
}

const availableSizes = {
  xs: 10,
  s: 12,
  m: 13,
  l: 14,
  xl: 18,
  xxl: 18,
  giant: 24,
}

/**
 * Text will be used for everywhere a text appears.
 * The only exception is in molecules that already provide the
 * text component for you.
 */
class Text extends React.Component {
  static propTypes = {
    align: PropTypes.oneOf(['left', 'center', 'right']),
    autoBreak: PropTypes.bool,
    block: PropTypes.bool,
    children: PropTypes.node,
    color: PropTypes.string,
    italic: PropTypes.bool,
    size: PropTypes.oneOf(['xs', 's', 'm', 'l', 'xl', 'giant']),
    strong: PropTypes.bool,
    underline: PropTypes.bool,
    lineThrough: PropTypes.bool,
  }

  static defaultProps = {
    size: 'l',
    block: true,
    color: 'text',
  }

  render() {
    const {
      color,
      block,
      children,
      italic,
      size,
      strong,
      align,
      underline,
      autoBreak,
      lineThrough,
      ...props
    } = this.props

    const styles = css({
      display: block ? 'block' : 'inline',
      fontFamily: '"Open Sans", Helvetica, Arial, sans-serif',
      fontStyle: italic && 'italic',
      fontWeight: strong && '600',
      fontSize: availableSizes[size],
      textDecoration:
        (underline && 'underline') || (lineThrough && 'line-through'),
      textAlign: align,
      whiteSpace: autoBreak && 'pre-wrap',
      wordBreak: autoBreak && 'break-word',
      WebkitFontSmoothing: 'antialiased',
      MozOsxFontSmoothing: 'grayscale',
    })

    return (
      <Theme>
        {({ colorize }) => (
          <View {...css(styles, { color: colorize(color) })} {...props}>
            {children}
          </View>
        )}
      </Theme>
    )
  }
}

export default Text
