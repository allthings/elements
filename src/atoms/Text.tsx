import React from 'react'
import { css } from 'glamor'
import View, { IView } from '../atoms/View'
import Theme, { IThemeChildren } from '../behaviour/Theme'
import neue from 'neue'

if (typeof window !== `undefined`) {
  neue.load([
    {
      families: ['Open Sans:n4,n4i,n6,n6i'],
      css: '//fonts.googleapis.com/css?family=Open+Sans:400,400i,600,600i',
    },
  ])
}

type sizeType = 'xs' | 's' | 'm' | 'l' | 'xl' | 'xxl' | 'giant'

const availableSizes = {
  xs: 10,
  s: 12,
  m: 13,
  l: 14,
  xl: 18,
  xxl: 18,
  giant: 24,
}

interface ITextStyles {
  size: sizeType
  block?: boolean
  italic?: boolean
  strong?: boolean
  underline?: boolean
  lineThrough?: boolean
  align?: 'left' | 'center' | 'right'
  autoBreak?: boolean
}

export const createTextStyles = ({
  block = false,
  italic = false,
  strong = false,
  size = 'm',
  underline = false,
  lineThrough = false,
  align,
  autoBreak,
}: ITextStyles) => {
  return css({
    display: block ? 'block' : 'inline',
    fontFamily: '"Open Sans", Helvetica, Arial, sans-serif',
    fontStyle: italic && 'italic',
    fontWeight: strong && '600',
    fontSize: availableSizes[size],
    textDecoration:
      (underline && 'underline') || (lineThrough && 'line-through'),
    textAlign: align,
    whiteSpace: autoBreak && 'pre-wrap',
    wordBreak: autoBreak && 'break-word',
    WebkitFontSmoothing: 'antialiased',
    MozOsxFontSmoothing: 'grayscale',
  })
}

export interface IText {
  color: string
}

/**
 * Text will be used for everywhere a text appears.
 * The only exception is in molecules that already provide the
 * text component for you.
 *
 * ```example
 * <Text>
 *   Text will be used for everywhere a text appears. The only exception
 *   is in molecules that already provide the text component for you.
 * </Text>
 * ```
 *
 * ```example
 * <Text autoBreak={true}>
 *   Textwillbeusedforeverywhereatextappears.Theonlyexceptionisinmoleculesthatalreadyprovidethetextcomponentforyou.Textwillbeusedforeverywhereatextappears.Theonlyexceptionisinmoleculesthatalreadyprovidethetextcomponentforyou.
 *  </Text>
 *  ```
 */
const Text = ({
  color = 'text',
  block = true,
  children,
  italic,
  size = 'l',
  strong,
  align,
  underline,
  autoBreak,
  lineThrough,
  ...props
}: IText & ITextStyles & IView) => (
  <Theme>
    {({ colorize }: IThemeChildren) => (
      <View
        {...css(
          createTextStyles({
            block,
            italic,
            strong,
            size,
            underline,
            lineThrough,
            align,
            autoBreak,
          }),
          { color: colorize(color) }
        )}
        {...props}
      >
        {children}
      </View>
    )}
  </Theme>
)

export default Text
