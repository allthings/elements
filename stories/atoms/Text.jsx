import React from 'react'
import Text from '../../src/atoms/Text'
import View from '@allthings/react-view'
import Example from '../Example'
import Notes from '../Notes'

class TextStory extends React.Component {
  render() {
    return (
      <View>
        <Notes for={Text} />
        <Example>
          <Text>
            Text will be used for everywhere a text appears. The only exception
            is in molecules that already provide the text component for you.
          </Text>
        </Example>
        <h3>Auto Break Example</h3>
        <Example>
          <Text autoBreak={true}>
            Textwillbeusedforeverywhereatextappears.Theonlyexceptionisinmoleculesthatalreadyprovidethetextcomponentforyou.
          </Text>
        </Example>
      </View>
    )
  }
}

export default TextStory
