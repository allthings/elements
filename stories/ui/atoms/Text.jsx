import React from 'react'
import Text from 'components/Text'
import View from '@allthings/react-view'
import Example from '../../Example'
import Notes from '../../Notes'

class TextStory extends React.Component {
  render() {
    return (
      <View>
        <Notes for={Text} />
        <Example>
          <Text>
            Text will be used for everywhere a text appears. The only exception
            is in molecules that already provide the text component for you.
          </Text>
        </Example>
        <h3>Auto Break Example</h3>
        <Example>
          <Text autoBreak={true}>
            Textwillbeusedforeverywhereatextappears.Theonlyexceptionisinmoleculesthatalreadyprovidethetextcomponentforyou.
          </Text>
        </Example>
      </View>
    )
  }
}

export default TextStory

/**
 import React from 'react'
 import Text from 'components/Text'
 import Card, { CardContent, Spacer } from 'components/Card'
 import { ColorPalette } from '@allthings/colors'
 import View from '@allthings/react-view'

 class TextStory extends React.Component {
  static propTypes = {}

  static defaultProps = {}

  constructor (props) {
    super(props)
  }

  render () {
    return (
      <View>
        <Card>
          <CardContent>
            <Text strong size="xxl">Text</Text>
            <Spacer />
            <Text>
              Text will be used for everywhere a text appears.
              The only exception is in molecules that already provide the
              text component for you.
            </Text>
          </CardContent>
        </Card>
        <Spacer />
        <Card>
          <CardContent>
            <Text strong size="l">Colors</Text>
            <Text>
              Colors are passed as a string. Available colors are:
            </Text>
            {Object.keys(ColorPalette.text).map(color => (
              <Text color={color} block={false}>{color}, </Text>))}
          </CardContent>
        </Card>
        <Spacer />
        <Card>
          <CardContent>
            <Text strong size="l">Block</Text>
            <Spacer />
            <Text>
              Texts are <Text block={false} strong>block by default</Text>, so
              make sure
              to use block={false} to set it non blocking, e.g. when use nested
            </Text>
            <Spacer />
            <Text strong size="l">Auto Break</Text>
            <Spacer />
            <Text>
              Texts can <Text block={false} strong>auto-break for you</Text>, so
              make sure
              to use autoBreak=true if you dont want text that freaks out.
            </Text>
            <Spacer />
            <Text strong size="m">Without Auto Break</Text>
            <Text>
              Loremipsumdolorsitametconsectetueradipiscingelit.
            </Text>
            <Spacer />
            <Text strong size="m">With Auto Break</Text>
            <Text autoBreak={true}>
              Loremipsumdolorsitametconsectetueradipiscingelit.
            </Text>
          </CardContent>
        </Card>
      </View>
    )
  }
}

 export default TextStory
 **/
