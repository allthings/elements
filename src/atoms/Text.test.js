import React from 'react'
import renderer from 'react-test-renderer'
import Text from './Text'

test('Text renders correctly', () => {
  const tree = renderer.create(
    <Text />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
