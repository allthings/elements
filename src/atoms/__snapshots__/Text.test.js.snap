// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text renders correctly 1`] = `
.css-1pkbzrh,
[data-css-1pkbzrh] {
  display: block;
  font-family: "Open Sans", Helvetica, Arial, sans-serif;
  font-size: 14px;
  color: #333333;
}

.css-4p3ux5,
[data-css-4p3ux5] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

<div
  data-css-1pkbzrh=""
  data-css-4p3ux5=""
/>
`;
